docs/proposals/202104_native_support_of_multitenancy.md

```diff
@@ -0,0 +1,83 @@
+
+### Summary
+
+We want to introduce a special tenant label for identifying a tenant and make each component aware of it and then finally track the data usages across the tenants. The receiver is on the safe side as it has a tenant-specific label to uniquely identify tenants.
+
+All the components implementing either the Store API (Sidecar, Ruler, Store Gateway, Receiver) or Query API (Thanos querier frontend and Thanos querier) will be affected. Currently, the ruler is exposing the store API but since there is an ongoing LFX project (to make the ruler stateless), it will upload the data to the receiver itself and it might no longer expose Store API.
+
+### Motivation
+
+* We actually don't have any way to track the resource usages of the query(like error rates, memory allocation, CPU usages along with the queries) and tenant-wise metrics.
+* Every component should allow passing a special label that allows APIs that identify tenants.
+* Currently, we use external labels to achieve multi-tenancy. These labels are sufficient for the sidecar (single-tenant per sidecar i.e hard tenant) but for other components (where we will have multiple tenants per component i.e soft tenancy) these external labels are not sufficient.
+
+### Goals
+
+1. Introduce a mechanism to identify tenants to support multi-tenancy use cases.
+2. Isolate all the incoming query requests.
+3. Track data usage across tenants.
+
+#### Use Case
+
+1. Allow admins to obtain tenant information on per tenant queries, operations, and ingestion.
+2. Sending tenant-specific alerts to different alert managers.
+3. Support cross tenant views.
+
+### Non Goals
+
+*  Not introduce any type of authorization/authenticatication mechanism.
+    * We should deal with auth outside Thanos using some authentication proxy (eg. kube-rbac-proxy) allowing us to use any authN/authZ system we want.
+    * Evaluate some set of rules and check if a certain label can access the data or not.
+
+### Solution
+
+* Introduce a new flag to enable or disable the multi-tenancy mode.
+* When this flag is enabled we will modify the metrics which can be used to track usage characteristics of a tenant by injecting the tenant’s label (a const. Label ).
+* We will make sure that the query request received has a tenant label and if not we have some default tenant label configured.
+* We will enforce the tenant labels inside the Thanos code base(i.e we assume the authentication is done and the query has access to the tenant's data which it specifies) when we run components with a specified mode(i.e multi-tenancy flag is enabled). We will no longer require any label enforcing proxy.
+* Support the cross tenant queries using regular expressions.
+* Store API will advertise the tenant label and this how it will help us -
+    * Let's assume a case of hard tenancy, we have two store-gateway advertising different sets of tenants.
```

Comment:
Do we plan to have a special Info method field in proto. If yes, do you think we could design it here in this doc?

Alternatively, why not reusing `LabelSet`?  (I am not suggesting we should, but it would be nice to mention why not e.g in `Alternatives` section below)

Reply:
Is this suggestion about this proposal - https://thanos.io/tip/proposals/202101_endpoint_discovery.md/

Reply:
Correct! cc @hitanshu-mehta